graph/graph.go

```diff
@@ -402,6 +404,43 @@ func (graph *Graph) ByParent() (map[string][]*image.Image, error) {
 	return byParent, err
 }
 
+// If the images and layers are in pulling chain, mark them.
+// If not, they may be deleted by rmi with dangling condition.
+func (graph *Graph) AddtoPulling(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.pullingImgs[layerID]; exists {
+			graph.pullingImgs[layerID] = append(imgs, imgID)
+		} else {
+			graph.pullingImgs[layerID] = []string{imgID}
+		}
+	}
+}
+
+// If the images and layers have benn pulled, unmark them.
+func (graph *Graph) DelfromPulling(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.pullingImgs[layerID]; exists {
+			result := make([]string, 0)
+			index := 0
+			endIndex := len(imgs) - 1
+			for k, v := range imgs {
```

Comment:
Why use a slice here when this is a membership operation? A map may be much more appropriate. This makes the code more complicated for a small, theoretical boost in performance.